src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -131,33 +122,43 @@ where s.Value.SagaData.GetType() == sagaType && (s.Key != saga.Id)
 
         ConcurrentDictionary<Guid, VersionedSagaEntity> data = new ConcurrentDictionary<Guid, VersionedSagaEntity>();
 
-        int version;
-
         class VersionedSagaEntity
         {
             public VersionedSagaEntity(IContainSagaData sagaData, VersionedSagaEntity original = null)
             {
                 SagaData = DeepClone(sagaData);
-                versionCache = original?.versionCache ?? new ConditionalWeakTable<IContainSagaData, SagaVersion>();
+                if (original != null)
+                {
+                    original.ConcurrencyCheck(sagaData);
+
+                    versionCache = original.versionCache;
+                    version = original.version;
+                    version++;
```

Comment:
didn't went too deep, if you're saying there is no risk in having multiple `VersionedSagaEntity` instances with the same id but different saga data, then I'm good with that :+1: